### PR TITLE
[fix bug 1383305] Add Fx55 WNP experiment.

### PR DESCRIPTION
**Pull request by jpetto**
## Description

Adds an experiment for locales *not* in the following list:

`('de', 'en-GB', 'en-US', 'en-ZA', 'es-ES', 'fr', 'id', 'ja', 'ja-JP-mac', 'pl', 'pt-BR', 'ru', 'zh-CN', 'zh-TW')`

`?v=a`: control
`?v=b`: QR code displayed, no app store badges displayed
`?v=c`: QR code displayed, app store badges moved up

Locales in the exclusion list should get the default behavior for that locale, carried over from WNP for Fx54.

This PR also ensures the country-specific store badge links work when the Send To Device widget is absent. (Previously this functionality piggy-backed on the Send To Device geolocation lookup.)

## Issue / Bugzilla link

https://bugzilla.mozilla.org/show_bug.cgi?id=1383305

## Testing

Make sure DNT is disabled to be entered in to the experiment. Can also hit the variations directly:

- https://bedrock-demo-jpetto.us-west.moz.works/it/firefox/55.0/whatsnew/?v=a (control)
- https://bedrock-demo-jpetto.us-west.moz.works/it/firefox/55.0/whatsnew/?v=b (only QR code)
- https://bedrock-demo-jpetto.us-west.moz.works/it/firefox/55.0/whatsnew/?v=c (QR code and store badges)

`?v=c` has custom adjust links, which look almost the same as the existing, except have `_v2` appended to the `&creative=` query param. The demo has a hack to spoof geolocation to test these  alternate links:

- https://bedrock-demo-jpetto.us-west.moz.works/it/firefox/55.0/whatsnew/?v=a&country=it (should see `&creative=badge_it` in store badge URL)
- https://bedrock-demo-jpetto.us-west.moz.works/it/firefox/55.0/whatsnew/?v=c&country=it (should see `&creative=badge_it_v2` in store badge URL)

### CODE REVIEW

File: bedrock/firefox/views.py [hunk 1]
```diff
@@ -406,7 +407,16 @@ def get_template_names(self):
...
-class WhatsnewView(l10n_utils.LangFilesMixin, TemplateView):
+class WhatsnewView(VariationMixin, l10n_utils.LangFilesMixin, TemplateView):
+    template_context_variations = ['a', 'b', 'c']
+    # locales *not* targeted by experiment
+    excluded_locales = ('de', 'en-GB', 'en-US', 'en-ZA', 'es-ES', 'fr', 'id',
+                        'ja', 'ja-JP-mac', 'pl', 'pt-BR', 'ru', 'zh-CN',
+                        'zh-TW')
+    # calculate non-excluded locales
+    # intent is to find locales in PROD_LANGUAGES *not* present in excluded_locales
+    variation_locales = list(set(settings.PROD_LANGUAGES).symmetric_difference(set(excluded_locales)))
```
Comment (jpetto) [inline bedrock/firefox/views.py]: @pmac Is there a more sane way to perform this operation?

Comment (pmac) [inline bedrock/firefox/views.py]: This will work, but it's a bit more confusing than what you really want I think. The following should be sufficient. 

```python
excluded_locales = {'de', 'en-GB', 'en-US', 'en-ZA', 'es-ES', 'fr', 'id',
                    'ja', 'ja-JP-mac', 'pl', 'pt-BR', 'ru', 'zh-CN',
                    'zh-TW'}
variation_locales = list(set(settings.PROD_LANGUAGES) - excluded_locales)
```

Note the curly braces around the `excluded_locales` value, that's a set literal, so it's already a set.

Comment (pmac) [inline bedrock/firefox/views.py]: Or of course:

```python
variation_locales = list(set(settings.PROD_LANGUAGES) - {
    'de', 'en-GB', 'en-US', 'en-ZA', 'es-ES', 'fr', 'id', 'ja',
    'ja-JP-mac', 'pl', 'pt-BR', 'ru', 'zh-CN', 'zh-TW',
})
```

Comment (jpetto) [inline bedrock/firefox/views.py]: Thanks @pmac!

File: bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html [hunk 2]
```diff
@@ -40,8 +46,15 @@
 {% set show_send_to_device = LANG in settings.SEND_TO_DEVICE_LOCALES %}
 
 {% set send_to_device_message_set = 'fx-50-whatsnew' %}
-{% set app_store_url = 'https://app.adjust.com/2uo1qc?campaign=whats_new&amp;adgroup=iOS_App_Store&amp;creative=badge_en&amp;fallback=https%3A%2F%2Fitunes.apple.com%2Fapp%2Fapple-store%2Fid989804926%3Fpt%3D373246%26ct%3Dadjust_tracker%26mt%3D8' %}
-{% set play_store_url = 'https://app.adjust.com/2uo1qc?campaign=whats_new&amp;adgroup=Google_Play_Store&amp;creative=badge_en' %}
+
+{% if variation == 'c' %}
+  {% set app_store_url = 'https://app.adjust.com/2uo1qc?campaign=whats_new&adgroup=iOS_App_Store&creative=badge_en_v2&fallback=https%3A%2F%2Fitunes.apple.com%2Fapp%2Fapple-store%2Fid989804926%3Fpt%3D373246%26ct%3Dadjust_tracker%26mt%3D8' %}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html]: Encode `&amp;`'s here and below?

File: bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html [hunk 3]
```diff
@@ -122,5 +149,13 @@ <h4 class="up-to-date"><span>{{ _('Your Firefox is up to date.') }}</span></h4>
 {% block email_form %}{% endblock %}
 
 {% block js %}
-  {% javascript 'firefox_whatsnew_54' %}
+  {% if variation == 'c' %}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html]: I don't suggest changing anything as this is all working, but fwiw I find all this logic in a single template a bit hard to follow at times. This is probably just personal preference, but I find separate template variations feel a bit cleaner.

File: media/js/firefox/whatsnew/whatsnew-54.js [hunk 4]
```diff
@@ -2,58 +2,106 @@
...
+                    }
+
+                    break;
+                }
             }
+
+            return newURL;
+        },
+        init: function() {
+            var xhr = new XMLHttpRequest();
```
Comment (alexgibson) [inline media/js/firefox/whatsnew/whatsnew-54.js]: Totally up to you and something I haven't tried using yet, but seeing as this is for FF 54, could you use `fetch` here instead of `XMLHttpRequest`? https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

File: bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html [hunk 5]
```diff
@@ -12,6 +12,12 @@
 
 {% extends "firefox/base-resp.html" %}
 
+{% block experiments %}
+  {% if switch('firefox-experiment-whatsnew-55', variation_locales) %}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html]: Again a switch here for `whatsnew-55` in the `whatsnew-45` template? I'm sure this is intentional, but just checking?

File: media/js/firefox/whatsnew/whatsnew-54.js [hunk 6]
```diff
@@ -2,58 +2,106 @@
...
+                    } catch (e) {
+                        country = 'none';
+                    }
+
+                    WNP54.updateStoreLinks(country);
+                }
+            };
+
+            xhr.open('GET', '/country-code.json');
+            // must come after open call above for IE 10 & 11
```
Comment (alexgibson) [inline media/js/firefox/whatsnew/whatsnew-54.js]: We're not worried about IE here, right?

File: bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html (hunk 5 above)
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html]: There are more things like this, such as images etc... but I'm guessing it's all as expected?

File: bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html (hunk 2 above)
Comment (jpetto) [inline bedrock/firefox/templates/firefox/whatsnew/fx54/whatsnew-54.html]: I'm not sure why, but when set with these variables, the ampersands never get decoded. If you copy this link and put it directly in an `href`, however, the ampersands *do* get decoded. 🤷‍♂️

File: media/js/firefox/whatsnew/whatsnew-54.js (hunk 4 above)
Comment (jpetto) [inline media/js/firefox/whatsnew/whatsnew-54.js]: Ah, rats! I *have* used the `fetch` API before, but was too concerned about finishing to think of it. Next time!

File: media/js/firefox/whatsnew/whatsnew-54.js (hunk 6 above)
Comment (jpetto) [inline media/js/firefox/whatsnew/whatsnew-54.js]: Nope, just a copy pasta comment.

File: media/js/firefox/whatsnew/whatsnew-54.js [hunk 7]
```diff
@@ -2,58 +2,105 @@
...
+            'jp': 'https://app.adjust.com/2uo1qc?campaign=whats_new&adgroup=Google_Play_Store&creative=badge_jp',
+            'mx': 'https://app.adjust.com/2uo1qc?campaign=whats_new&adgroup=Google_Play_Store&creative=badge_es-mx',
+            'pl': 'https://app.adjust.com/2uo1qc?campaign=whats_new&adgroup=Google_Play_Store&creative=badge_pl',
+            'ru': 'https://app.adjust.com/2uo1qc?campaign=whats_new&adgroup=Google_Play_Store&creative=badge_ru'
+        },
+        updateStoreLinks: function(country) {
+            var newAppStoreURL = WNP54.getStoreURL(country, 'ios');
+            var newPlayStoreURL = WNP54.getStoreURL(country, 'android');
 
-    var form = new Mozilla.SendToDevice();
```
Comment (alexgibson) [inline media/js/firefox/whatsnew/whatsnew-54.js]: Hmm, isn't this needed for the regular whatsnew 54 page? The AJAX on the form no longer seems to work.

File: media/js/firefox/whatsnew/whatsnew-54.js [hunk 8]
```diff
@@ -83,29 +83,22 @@ if (typeof window.Mozilla === 'undefined') {
...
-                }
-            };
-
-            xhr.open('GET', '/country-code.json');
-            xhr.timeout = 2000;
-            xhr.send();
+                form.init();
+            } else {
+                // make the request
+                fetch('/country-code.json', {
```
Comment (pmac) [inline media/js/firefox/whatsnew/whatsnew-54.js]: 👍

File: bedrock/firefox/views.py [hunk 9]
```diff
@@ -407,6 +407,14 @@ def get_template_names(self):
 
 
 class WhatsnewView(l10n_utils.LangFilesMixin, TemplateView):
+    # locales *not* targeted by experiment
+    excluded_locales = {'de', 'en-GB', 'en-US', 'en-ZA', 'es-ES', 'fr', 'id',
+                        'ja', 'ja-JP-mac', 'pl', 'pt-BR', 'ru', 'zh-CN',
+                        'zh-TW'}
+
+    # calculate non-excluded locales
+    variation_locales = list(set(settings.PROD_LANGUAGES) - excluded_locales)
```
Comment (pmac) [inline bedrock/firefox/views.py]: 👍

File: media/js/firefox/whatsnew/whatsnew-54.js [hunk 10]
```diff
@@ -2,58 +2,105 @@
...
+                form.geoCallback = WNP54.updateStoreLinks;
+
+                form.init();
+            } else {
+                // make the request
+                fetch('/country-code.json', {
+                    method: 'GET'
+                }).then(function(response) {
+                    // convert response to JSON
+                    return response.json();
```
Comment (alexgibson) [inline media/js/firefox/whatsnew/whatsnew-54.js]: Hmm, do we need to add a check here that the `fetch` was successful? The MDN docs seem to say that something like a 404 will not throw an error, so it's best to check that `response.ok` is true. There's also an example of checking for a valid content type of `application/json`.

https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

Comment (jpetto) [inline media/js/firefox/whatsnew/whatsnew-54.js]: Good call on making sure we have a response. I think checking for content type is overkill, as we control the service side and have tests.
